opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py

```diff
@@ -349,6 +362,326 @@ def collect(
+                bucket_counts=self._negative.counts,
+            ),
+            # FIXME: Find the right value for flags
+            flags=0,
+            min=min_,
+            max=max_,
+        )
+
+        if self._previous_point is None or (
+            self._instrument_temporality is aggregation_temporality
+        ):
+            self._previous_point = current_point
+            return current_point
+
+        max_ = current_point.max
+        min_ = current_point.min
+
+        if aggregation_temporality is AggregationTemporality.CUMULATIVE:
+            start_time_unix_nano = self._previous_point.start_time_unix_nano
+            sum_ = current_point.sum + self._previous_point.sum
+            # Only update min/max on delta -> cumulative
+            max_ = max(current_point.max, self._previous_point.max)
+            min_ = min(current_point.min, self._previous_point.min)
+
+            negative_counts = [
+                curr_count + prev_count
+                for curr_count, prev_count in zip(
+                    current_point.negative.bucket_counts,
+                    self._previous_point.negative.bucket_counts,
+                )
+            ]
+            positive_counts = [
+                curr_count + prev_count
+                for curr_count, prev_count in zip(
+                    current_point.positive.bucket_counts,
+                    self._previous_point.positive.bucket_counts,
+                )
+            ]
+        else:
+            start_time_unix_nano = self._previous_point.time_unix_nano
+            sum_ = current_point.sum - self._previous_point.sum
+
+            negative_counts = [
+                curr_count - prev_count
+                for curr_count, prev_count in zip(
+                    current_point.negative.bucket_counts,
+                    self._previous_point.negative.bucket_counts,
+                )
+            ]
+            positive_counts = [
+                curr_count - prev_count
+                for curr_count, prev_count in zip(
+                    current_point.positive.bucket_counts,
+                    self._previous_point.positive.bucket_counts,
+                )
+            ]
+
+        current_point = ExponentialHistogramDataPoint(
+            attributes=self._attributes,
+            start_time_unix_nano=start_time_unix_nano,
+            time_unix_nano=current_point.time_unix_nano,
+            count=self._count,
+            sum=sum_,
+            scale=scale,
+            zero_count=self._zero_count,
+            positive=BucketsPoint(
+                offset=self._positive.offset, bucket_counts=positive_counts
+            ),
+            negative=BucketsPoint(
+                offset=self._negative.offset, bucket_counts=negative_counts
+            ),
+            # FIXME: Find the right value for flags
+            flags=0,
+            min=min_,
+            max=max_,
+        )
+
+        self._previous_point = current_point
+
+        with self._lock:
```

Comment:
It seems like all this "self.X" stuff should be ABOVE in the same first lock.

I'd recommend copying all the data to local variables *or* keeping the lock on this entire method here if you want it to be accurate.

I thought that's how you'd done the other aggregations, but as @aabmass called out, this may not be the case and we need to reevaluate the other ones too.

Reply:
Added the lock for the whole method.